File: perl/worker.pl
#!/usr/bin/perl

use strict;
use warnings;

use Net::RabbitFoot;

my $conn = Net::RabbitFoot->new()->load_xml_spec()->connect(
    host => 'localhost',
    port => 5672,
    user => 'guest',
    pass => 'guest',
    vhost => '/',
);

my $ch = $conn->open_channel();

$ch->declare_queue(
    queue => 'task_queue',
    durable => 1,
);

print " [*] Waiting for messages. To exit press CTRL-C\n";

sub callback {
    my $var = shift;
    my $body = $var->{body}->{payload};
    print " [x] Received $body\n";

    my @c = $body =~ /\./g;
    sleep(scalar(@c));

    print " [x] Done\n";
    $ch->ack();
}

$ch->qos(prefetch_count => 1,);

$ch->consume(
    on_consume => \&callback,
    no_ack => 0,
);

# Wait forever
AnyEvent->condvar->recv;

File: dart/worker.dart
import "dart:io";
import "package:dart_amqp/dart_amqp.dart";

void main (List<String> arguments) {
    ConnectionSettings settings = new ConnectionSettings(
        host: "localhost"
    );

    Client client = new Client(settings: settings);

    ProcessSignal.sigint.watch().listen((_) {
        client.close().then((_) {
            print("close client");
            exit(0);
        });
    });

    String consumeTag = "task_queue";

    client
        .channel()
        .then((Channel channel) {
            return channel.qos(0, 1)
                .then((Channel channel) =>
                                channel.queue(consumeTag, durable: true));
        })
        .then((Queue queue) => queue.consume(noAck: false))
        .then((Consumer consumer) {
            consumer.listen((AmqpMessage event) {
                String payload = event.payloadAsString;
                print(" [x] Received ${payload}");
                sleep(new Duration(seconds : payload.split(".").length));
                print(" [x] Done");
                event.ack();
            });
        });
}
